In [1]:
import pandas as pd
import requests

from accessibility import check_endpoint
from summarize import * #json_keys, data_frame

#### ARGO data API

API points to: "ARGO Floats data"

API documentation (swagger): https://dataselection.euro-argo.eu/swagger-ui.html#/

##### Basins

In [2]:
# Analysed endpoint (selected from API docs)
argo_data_url = "https://dataselection.euro-argo.eu/api/basins-tree"

In [3]:
endpoint_response = requests.get(argo_data_url).json()

In [4]:
summary = json_keys(endpoint_response)

for key, info in summary.items():
    print(f"Key: {key}, Count: {info['count']}, Types: {', '.join(info['types'])}, Example value: '{info['example']}'")

Key: id, Count: 11, Types: int, Example value: '23617'
Key: name, Count: 11, Types: str, Example value: 'ARCTIC OCEAN'
Key: subBasins, Count: 11, Types: list, Example value: '[{'id': 24027, 'name': 'KARA SEA'}, {'id': 24029, 'name': 'BARENTS SEA'}, {'id': 24026, 'name': 'LAPTEV SEA'}, {'id': 24020, 'name': 'WHITE SEA'}, {'id': 24021, 'name': 'ICELAND SEA'}, {'id': 24024, 'name': 'NORWEGIAN SEA'}, {'id': 24025, 'name': 'EAST SIBERIAN SEA'}, {'id': 24028, 'name': 'NORTHWESTERN PASSAGES'}, {'id': 24019, 'name': 'DAVIS STRAIT'}, {'id': 24030, 'name': 'BAFFIN BAY'}, {'id': 24032, 'name': 'LINCOLN SEA'}, {'id': 24018, 'name': 'HUDSON BAY'}, {'id': 24023, 'name': 'BEAUFORT SEA'}, {'id': 36279, 'name': 'GREENLAND SEA'}, {'id': 24017, 'name': 'HUDSON STRAIT'}, {'id': 24022, 'name': 'CHUKCHI SEA'}, {'id': 26579, 'name': 'FRAM STRAIT'}]'
Key: subBasins.id, Count: 236, Types: int, Example value: '26579'
Key: subBasins.name, Count: 236, Types: str, Example value: 'FRAM STRAIT'


In [5]:
# Convert the summary dictionary to a pandas DataFrame
df_summ= pd.DataFrame([
    {
        "Property": key,
        "Count": info["count"],
        "Types": ', '.join(info["types"]),
        "Example": info["example"]
    }
    for key, info in summary.items()
])
df_summ.to_csv("properties/ARGO_JSONAPI_metadata_basins.csv", index=False)

This endpoint returns a list of ocean basins and their sub-basins,  
this can be visualized as:  
![image.png](images/ARGO_JSONAPI_oceanbasins.drawio.png)

##### Cycle by cycle-id

In [36]:
# Analysed endpoint (selected from API docs)
argo_data_url = "https://dataselection.euro-argo.eu/api/find-by-id/760094"

In [37]:
endpoint_response = requests.get(argo_data_url).json()

In [38]:
summary = json_keys(endpoint_response)

for key, info in summary.items():
    print(f"Key: {key}, Count: {info['count']}, Types: {', '.join(info['types'])}, Example value: '{info['example']}'")

Key: id, Count: 1, Types: int, Example value: '760094'
Key: cvNumber, Count: 1, Types: int, Example value: '1'
Key: startDate, Count: 1, Types: str, Example value: '2008-03-23T02:00:54.000+0000'
Key: endDate, Count: 1, Types: str, Example value: '2008-03-23T02:00:54.000+0000'
Key: coordinate, Count: 1, Types: dict, Example value: '{'lat': -28.184, 'lon': -22.997}'
Key: coordinate.lat, Count: 1, Types: float, Example value: '-28.184'
Key: coordinate.lon, Count: 1, Types: float, Example value: '-22.997'
Key: globalGeoShapeField, Count: 1, Types: dict, Example value: '{'x': -22.997, 'y': -28.184, 'coordinates': [-22.997, -28.184], 'type': 'Point'}'
Key: globalGeoShapeField.x, Count: 1, Types: float, Example value: '-22.997'
Key: globalGeoShapeField.y, Count: 1, Types: float, Example value: '-28.184'
Key: globalGeoShapeField.coordinates, Count: 1, Types: list, Example value: '[-22.997, -28.184]'
Key: globalGeoShapeField.type, Count: 1, Types: str, Example value: 'Point'
Key: platformCode, 

In [39]:
# Convert the summary dictionary to a pandas DataFrame
df_summ= pd.DataFrame([
    {
        "Property": key,
        "Count": info["count"],
        "Types": ', '.join(info["types"]),
        "Example": info["example"]
    }
    for key, info in summary.items()
])
df_summ.to_csv("properties/ARGO_JSONAPI_metadata_cycle.csv", index=False)

The endpoint returns metadata associated with the given cycle
this can be visualized as follows: 
![image.png](images/ARGO_JSONAPI_cycle.drawio.png)

##### Cycle by platformID and cvNumber

In [40]:
# Analysed endpoint (selected from API docs)
argo_data_url = "https://dataselection.euro-argo.eu/api/find-by-platformid/3900675/cvnumber/1"

In [41]:
endpoint_response = requests.get(argo_data_url).json()

In [42]:
summary = json_keys(endpoint_response)

for key, info in summary.items():
    print(f"Key: {key}, Count: {info['count']}, Types: {', '.join(info['types'])}, Example value: '{info['example']}'")

Key: id, Count: 1, Types: int, Example value: '760094'
Key: cvNumber, Count: 1, Types: int, Example value: '1'
Key: startDate, Count: 1, Types: str, Example value: '2008-03-23T02:00:54.000+0000'
Key: endDate, Count: 1, Types: str, Example value: '2008-03-23T02:00:54.000+0000'
Key: coordinate, Count: 1, Types: dict, Example value: '{'lat': -28.184, 'lon': -22.997}'
Key: coordinate.lat, Count: 1, Types: float, Example value: '-28.184'
Key: coordinate.lon, Count: 1, Types: float, Example value: '-22.997'
Key: globalGeoShapeField, Count: 1, Types: dict, Example value: '{'x': -22.997, 'y': -28.184, 'coordinates': [-22.997, -28.184], 'type': 'Point'}'
Key: globalGeoShapeField.x, Count: 1, Types: float, Example value: '-22.997'
Key: globalGeoShapeField.y, Count: 1, Types: float, Example value: '-28.184'
Key: globalGeoShapeField.coordinates, Count: 1, Types: list, Example value: '[-22.997, -28.184]'
Key: globalGeoShapeField.type, Count: 1, Types: str, Example value: 'Point'
Key: platformCode, 

Endpoint returns the metadata associated with a cycle that is defined by the given platformcode and cvNumber  
This metadata consists of the same properties as the metadata returned by the previous endpoint (https://dataselection.euro-argo.eu/api/find-by-id/{cycleID})  
And is visualized by the same diagram as above

No other endpoints, that allow Get requests, in the API documentation

### Analysis results:

- both endpoints (data API and metadata API) are json APIs,
- data findable and machine accessible 
- requires knowledge on API structure/model in order to understand and use data  
but this information is available with the Swagger API documentation 

- some of the Swagger APi documentation is not clear to non-domain experts  
  - e.g. some api paths require knowledge on the wmo-codes, platform-codes, ... 
  - the semantics are not clear

- Data returned by APIs:
  - 19045 Floats:
    - described by 39 properties (many more when taking nested properties into account (see diagram for overview))

  - 11 Basins:
    - described by 3 properties
    - 236 sub-basins

  - 900+ Cycles:
    - described by 30 properties (51 taking nested properties into account)
    - no exact number on total number of cycles

- Semantic ambiguity:
  - use of string/number codes as identifiers
  - limited/no use of persistent identifiers in data
  - limited/no use of external standard terms
  - sometimes unclear ambiguity of properties 